### PR TITLE
First steps for py3 migration

### PR DESCRIPTION
**Pull request by decentral1se**
Things I'm unsure about:
  * How to handle the `debian/changelog` file version bump?
  * How to handle the `setup.py` version bump? It's out of sync with ^^^?

This is just one part of this migration. I'll need to manually test this installed against a live YunoHost and probably need to refactor the LDAP code (as far as I remember, the library bindings have changed quite a bit). Maybe I should work in another branch?

I ran:

```
$ 2to3 -f all -f idioms -wn .
```

And then took a manual pass through the code to clean up the rest.

Refs:
  * https://github.com/YunoHost/issues/issues/1255
  * https://github.com/YunoHost/issues/issues/1367

### CODE REVIEW

File: doc/ldif2dot-0.1.py [hunk 1]
```diff
@@ -200,4 +200,4 @@ def parse(self, file, name):
     else:
         name = '<stdin>'
         file = sys.stdin
-    print Converter().parse(file, name)
+    print((Converter().parse(file, name)))
```
Comment (Psycojoker) [inline doc/ldif2dot-0.1.py:203]: Too much ()

```suggestion
    print(Converter().parse(file, name))
```

File: generate_api_doc.py [hunk 2]
```diff
@@ -56,7 +55,7 @@ def main():
     resources = {}
 
     del action_map['general_arguments']
-    for category, category_params in action_map.items():
+    for category, category_params in list(action_map.items()):
```
Comment (Psycojoker) [inline generate_api_doc.py:58]: Do we really all those `list()`? My tests suggest that no since we are in a fort loop and the iterator pattern is going to apply.

File: moulinette/actionsmap.py [hunk 3]
```diff
@@ -276,7 +270,7 @@ def validate(self, arg_name, parameters):
 
         """
         # Iterate over parameters to validate
-        for p, v in parameters.items():
+        for p, v in list(parameters.items()):
```
Comment (Psycojoker) [inline moulinette/actionsmap.py:273]: Same for all those `list()` (the previous one for string formatting is legit)

File: moulinette/authenticators/ldap.py [hunk 4]
```diff
@@ -260,7 +256,7 @@ def get_conflict(self, value_dict, base_dn=None):
             None | list with Fist conflict attribute name and value
 
         """
-        for attr, value in value_dict.items():
+        for attr, value in list(value_dict.items()):
```
Comment (Psycojoker) [inline moulinette/authenticators/ldap.py:259]: Same.

File: moulinette/core.py [hunk 5]
```diff
@@ -215,7 +213,7 @@ def set_locale(self, locale):
         self.locale = locale
 
         self._global.set_locale(locale)
-        for n in self._namespaces.values():
+        for n in list(self._namespaces.values()):
```
Comment (Psycojoker) [inline moulinette/core.py:216]: Same.

File: moulinette/interfaces/__init__.py [hunk 6]
```diff
@@ -284,7 +282,7 @@ def _validate_conf(self, configuration, is_global=False):
                                    "for 'authenticator'")
                 else:
                     auths = {}
-                    for auth_name, auth_conf in auth.items():
+                    for auth_name, auth_conf in list(auth.items()):
```
Comment (Psycojoker) [inline moulinette/interfaces/__init__.py:285]: Same.

File: moulinette/interfaces/__init__.py [hunk 7]
```diff
@@ -600,11 +598,11 @@ def format_help(self):
                 subcategories_subparser = copy.copy(action_group._group_actions[0])
 
                 # Filter "action"-type and "subcategory"-type commands
-                actions_subparser.choices = OrderedDict([(k, v) for k, v in actions_subparser.choices.items() if v.type == "action"])
-                subcategories_subparser.choices = OrderedDict([(k, v) for k, v in subcategories_subparser.choices.items() if v.type == "subcategory"])
+                actions_subparser.choices = OrderedDict([(k, v) for k, v in list(actions_subparser.choices.items()) if v.type == "action"])
+                subcategories_subparser.choices = OrderedDict([(k, v) for k, v in list(subcategories_subparser.choices.items()) if v.type == "subcategory"])
```
Comment (Psycojoker) [inline moulinette/interfaces/__init__.py:602]: Same, iterator pattern again.

File: moulinette/interfaces/__init__.py [hunk 8]
```diff
@@ -600,11 +598,11 @@ def format_help(self):
...
 
-                actions_choices = actions_subparser.choices.keys()
-                subcategories_choices = subcategories_subparser.choices.keys()
+                actions_choices = list(actions_subparser.choices.keys())
+                subcategories_choices = list(subcategories_subparser.choices.keys())
```
Comment (Psycojoker) [inline moulinette/interfaces/__init__.py:605]: Those ones are legit.

File: moulinette/interfaces/api.py [hunk 9]
```diff
@@ -114,7 +112,7 @@ def get_default(self, dest):
         return self._parser.get_default(dest)
 
     def add_arguments(self, arguments, extraparser, format_arg_names=None, validate_extra=True):
-        for argument_name, argument_options in arguments.items():
+        for argument_name, argument_options in list(arguments.items()):
```
Comment (Psycojoker) [inline moulinette/interfaces/api.py:115]: Same.

File: moulinette/interfaces/api.py [hunk 10]
```diff
@@ -566,7 +564,7 @@ def __init__(self, parent=None, **kwargs):
     @property
     def routes(self):
         """Get current routes"""
-        return self._parsers.keys()
+        return list(self._parsers.keys())
```
Comment (Psycojoker) [inline moulinette/interfaces/api.py:567]: Legit.

File: moulinette/interfaces/cli.py [hunk 11]
```diff
@@ -86,12 +84,10 @@ def plain_print_dict(d, depth=0):
         for v in d:
             plain_print_dict(v, depth + 1)
     elif isinstance(d, dict):
-        for k, v in d.items():
-            print("{}{}".format("#" * (depth + 1), k))
+        for k, v in list(d.items()):
```
Comment (Psycojoker) [inline moulinette/interfaces/cli.py:87]: Same.

File: moulinette/interfaces/cli.py [hunk 12]
```diff
@@ -86,12 +84,10 @@ def plain_print_dict(d, depth=0):
...
+            print(("{}{}".format("#" * (depth + 1), k)))
```
Comment (Psycojoker) [inline moulinette/interfaces/cli.py:88]: Too much ()

```suggestion
            print("{}{}".format("#" * (depth + 1), k))
```

File: moulinette/interfaces/cli.py [hunk 13]
```diff
@@ -142,27 +138,23 @@ def pretty_print_dict(d, depth=0):
...
                 elif isinstance(value, dict):
                     pretty_print_dict({key: value}, depth + 1)
                 else:
-                    if isinstance(value, unicode):
-                        value = value.encode('utf-8')
-                    elif isinstance(v, date):
+                    if isinstance(v, date):
                         v = pretty_date(v)
-                    print("{:s}- {}".format("  " * (depth + 1), value))
+                    print(("{:s}- {}".format("  " * (depth + 1), value)))
```
Comment (Psycojoker) [inline moulinette/interfaces/cli.py:153]: Too much ()

```suggestion
                    print("{:s}- {}".format("  " * (depth + 1), value))
```

File: moulinette/interfaces/cli.py [hunk 14]
```diff
@@ -142,27 +138,23 @@ def pretty_print_dict(d, depth=0):
         if isinstance(v, list) and len(v) == 1:
             v = v[0]
         if isinstance(v, dict):
-            print("{:s}{}: ".format("  " * depth, k))
+            print(("{:s}{}: ".format("  " * depth, k)))
             pretty_print_dict(v, depth + 1)
         elif isinstance(v, list):
-            print("{:s}{}: ".format("  " * depth, k))
+            print(("{:s}{}: ".format("  " * depth, k)))
```
Comment (Psycojoker) [inline moulinette/interfaces/cli.py:144]: It's weird seing all those "too much ()" on every print :|

File: moulinette/interfaces/cli.py [hunk 15]
```diff
@@ -131,7 +127,7 @@ def pretty_print_dict(d, depth=0):
         - depth -- The recursive depth of the dictionary
 
     """
-    keys = d.keys()
+    keys = list(d.keys())
```
Comment (Psycojoker) [inline moulinette/interfaces/cli.py:130]: To be confirmed (I don't know if "keys" is accessed using []) but I don't think it's needed.

File: moulinette/interfaces/cli.py [hunk 16]
```diff
@@ -467,7 +459,7 @@ def _do_prompt(self, message, is_password, confirm, color='blue'):
             prompt = lambda m: getpass.getpass(colorize(m18n.g('colon', m),
                                                         color))
         else:
-            prompt = lambda m: raw_input(colorize(m18n.g('colon', m), color))
+            prompt = lambda m: eval(input(colorize(m18n.g('colon', m), color)))
```
Comment (Psycojoker) [inline moulinette/interfaces/cli.py:462]: Do we really need an `eval` here ಠಿ_ಠ ?

File: moulinette/utils/filesystem.py [hunk 17]
```diff
@@ -34,7 +34,7 @@ def read_file(file_path):
             file_content = f.read()
     except IOError as e:
         raise MoulinetteError('cannot_open_file', file=file_path, error=str(e))
-    except Exception:
+    except Exception as e:
```
Comment (Psycojoker) [inline moulinette/utils/filesystem.py:37]: Oh, this one looks like an error that need to be fixed in another PR.

**Comment by Psycojoker** [comment]
I've stopped commenting nearly the end since it's was always the same comment everywhere:

* we don't need those `list(dict.{items,keys,values}())` in every iterator pattern situation
* there are too much `()` for all the prints and they aren't needed

I believe it's worth removing all of those to make a way smaller and easier to read diff for the others '-'

Other comments are about:

* this weird `eval` which really looks like a security issue
* the smole exception fix that should be in another PR

**Comment by decentral1se** [comment]
Thanks for reviewing @Psycojoker. I didn't bother to try and "clean" what `2to3` gave me because I have my mind on the bigger fish to fry on the migration. The tests pass, so I was happy to see that the code can execute at least. But I see that you find it important, so I will come to this at some point.

**Comment by decentral1se** [comment]
I'll come back to this.
